webrisk/cache.go

```diff
@@ -0,0 +1,176 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webrisk
+
+import (
+	"sync"
+	"time"
+
+	pb "github.com/GoogleCloudPlatform/golang-samples/webrisk/internal/webrisk_proto"
+	pt "github.com/golang/protobuf/ptypes"
+)
+
+type cacheResult int
+
+const (
+	// positiveCacheHit indicates that the given hash matched an entry in the cache.
+	// The caller must consider the match a threat and not contact the server.
+	positiveCacheHit cacheResult = iota
+
+	// negativeCacheHit indicates that the given hash did not match any entries
+	// in the cache but its prefix matches the negative cache. The caller must
+	// consider the given hash to be safe and not contact the server.
+	negativeCacheHit
+
+	// cacheMiss indicates that the given hash did not match any entry
+	// in the cache. The caller should make a follow-up query to the server.
+	cacheMiss
+)
+
+// cache caches results from API calls to SearchHashesRequest to reduce
+// network calls for recently requested items. Since the global blacklist is
+// constantly changing, the Web Risk API defines TTLs for how long entries
+// can stay alive in the cache.
+type cache struct {
+	sync.RWMutex
+
+	// pttls maps full hashes and a ThreatType to a positive time-to-live.
+	// For a given full hash, the known threats are all ThreatTypes that
+	// map to valid TTLs (i.e. in the future).
+	pttls map[hashPrefix]map[ThreatType]time.Time
+
+	// nttls maps partial hashes to a negative time-to-live.
+	// If this is still valid (i.e. in the future), then this indicates that
+	// there are *no* threats under the given partial hash, unless there exist
+	// ThreatTypes with a valid positive TTL for that hash.
+	nttls map[hashPrefix]time.Time
+
+	now func() time.Time
+}
+
+// Update updates the cache according to the request that was made to the server
+// and the response given back.
+func (c *cache) Update(req *pb.SearchHashesRequest, resp *pb.SearchHashesResponse) {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.pttls == nil {
+		c.pttls = make(map[hashPrefix]map[ThreatType]time.Time)
+		c.nttls = make(map[hashPrefix]time.Time)
+	}
+
+	// Insert each threat match into the cache by full hash.
+	for _, threat := range resp.GetThreats() {
+		fullHash := hashPrefix(threat.Hash)
+		if !fullHash.IsFull() {
+			continue
+		}
+		if c.pttls[fullHash] == nil {
+			c.pttls[fullHash] = make(map[ThreatType]time.Time)
+		}
+		for _, tt := range threat.ThreatTypes {
+			var err error
+			c.pttls[fullHash][ThreatType(tt)], err = pt.Timestamp(threat.ExpireTime)
+			if err != nil {
+				panic("timestamp conversion error")
```

Comment:
Consider returning an error instead of panicing. Otherwise, update the function comment saying under what conditions this will panic.

Reply:
Done